lib/srv/desktop/tdp/conn.go

```diff
@@ -164,3 +168,167 @@ func IsFatalErr(err error) bool {
 
 	return !IsNonFatalErr(err)
 }
+
+// NewConnProxy creates a bidirectional proxy to copy messages between the client and server connection.
+// It accepts an optional serverInterceptor to intercept received messages.
+func NewConnProxy(client, server io.ReadWriteCloser, serverInterceptor ServerInterceptor) *ConnProxy {
+	return &ConnProxy{
+		client:            client,
+		server:            server,
+		serverInterceptor: serverInterceptor,
+		messagesToClient:  make(chan Message),
+	}
+}
+
+// ConnProxy does a bidirectional copy between the connection to the client and the mTLS connection to the server.
+type ConnProxy struct {
+	// client is a connection to the client (browser/Connect).
+	client io.ReadWriteCloser
+	// server io.ReadWriteCloser is a connection to the server (Windows Desktop Service).
+	server io.ReadWriteCloser
+	// serverInterceptor intercepts the incoming messages.
+	// If the returned message is non-nil, it is forwarded to the client.
+	// If an error is returned, the stream is canceled.
+	serverInterceptor ServerInterceptor
+	messagesToClient  chan Message
+}
+
+// ServerInterceptor intercepts messages received from the server.
+// If a message returned from the interceptor is nil, it's not sent to the client.
+type ServerInterceptor func(serverTdpConn *Conn, message Message) (Message, error)
+
+// SendToClient sends a message to the client and blocks until the operation completes.
+func (c *ConnProxy) SendToClient(ctx context.Context, message Message) error {
+	select {
+	case c.messagesToClient <- message:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		return trace.Errorf("could not send TDP message to client")
+	}
+}
+
+// Run starts proxying the connection.
+func (c *ConnProxy) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+
+	var closeOnce sync.Once
+	closeAll := func() {
+		cancel()
+		c.client.Close()
+		c.server.Close()
+	}
+	defer closeOnce.Do(closeAll)
+
+	var errs errgroup.Group
+
+	sendTDPAlert := func(err error, severity Severity) error {
+		msg := Alert{Message: err.Error(), Severity: severity}
+		b, err := msg.Encode()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		_, err = c.client.Write(b)
+		return trace.Wrap(err)
+	}
+
+	// Run a goroutine to pick TDP messages up from a channel and send
+	// them to the client.
+	errs.Go(func() error {
+		defer closeOnce.Do(closeAll)
+
+		for {
+			select {
+			case msg, ok := <-c.messagesToClient:
+				if !ok {
+					return nil // channel closed gracefully
```

Comment:
Is anything closing this channel?

Reply:
No, nothing closes it. I added it more for a safety check, but since we control the lifecycle of `c.messagesToClient` channel it's not needed.